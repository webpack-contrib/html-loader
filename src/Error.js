class LoaderError extends Error {
  constructor(err) {
    super(err);

    this.name = 'HTML Loader';
    this.message = `\n\n${this.name}\n\n`;

    if (err.name === 'AttributesError') {
      this.message += `[option.attrs] ${err.message}\n`;
    } else {
      this.message += `${err.message}\n`;
    }

    this.stack = false;
  }
}

export default LoaderError;
