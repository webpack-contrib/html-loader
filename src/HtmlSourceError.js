function getIndices(value) {
  const result = [];
  let index = value.indexOf("\n");

  while (index !== -1) {
    result.push(index + 1);
    index = value.indexOf("\n", index + 1);
  }

  result.push(value.length + 1);

  return result;
}

function offsetToPosition(source, offset) {
  let index = -1;
  const indices = getIndices(source);
  const { length } = indices;

  if (offset < 0) {
    return {};
  }

  // eslint-disable-next-line no-plusplus
  while (++index < length) {
    if (indices[index] > offset) {
      return {
        line: index + 1,
        column: offset - (indices[index - 1] || 0) + 1,
        offset,
      };
    }
  }

  return {};
}

export default class HtmlSourceError extends Error {
  constructor(error, startOffset, endOffset, source) {
    super(error);

    this.name = "HtmlSourceError";
    this.message = `${this.name}: ${this.message}`;
    this.startOffset = startOffset;
    this.endOffset = endOffset;
    this.source = source;

    const startPosition = offsetToPosition(source, this.startOffset);
    const endPosition = offsetToPosition(source, this.endOffset);

    this.message += ` (From line ${startPosition.line}, column ${startPosition.column}; to line ${endPosition.line}, column ${endPosition.column})`;

    // We don't need stack
    this.stack = false;
  }
}
