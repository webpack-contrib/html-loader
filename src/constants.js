export const GET_URL_CODE =
  'function __url__(url) { return url.__esModule ? url.default : url; }';

export const IDENT_REGEX = /xxxHTMLLINKxxx[0-9.]+xxx/g;

export const REQUIRE_REGEX = /\${require\([^)]*\)}/g;
