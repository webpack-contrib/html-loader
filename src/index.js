/* eslint-disable */
import { getOptions } from 'loader-utils';
import validateOptions from 'schema-utils';

import posthtml from 'posthtml';
import urls from './lib/plugins/url';
import imports from './lib/plugins/import';
import minifier from 'htmlnano';

import schema from './options.json';
import LoaderError from './lib/Error';

// Loader Defaults
const defaults = {
  url: true,
  import: true,
  minimize: false,
  template: false,
};

export default function loader(html, map, meta) {
  // Loader Options
  const options = Object.assign(defaults, getOptions(this));

  validateOptions(schema, options, 'HTML Loader');
  // Make the loader async
  const cb = this.async();
  const file = this.resourcePath;

  // HACK add Module.type
  this._module.type = 'text/html';

  const template = options.template
    ? typeof options.template === 'string' ? options.template : '_'
    : false;

  const plugins = [];

  if (options.url) plugins.push(urls());
  if (options.import) plugins.push(imports({ template }));
  // TODO(michael-ciniawsky)
  // <imports src=""./file.html"> aren't minified (#160)
  if (options.minimize) plugins.push(minifier());

  // Reuse HTML AST (PostHTML AST) if available
  // (e.g posthtml-loader) to avoid HTML reparsing
  if (meta && meta.ast && meta.ast.type === 'posthtml') {
    html = meta.ast.root;
  }

  posthtml(plugins)
    .process(html, { from: file, to: file })
    .then(({ html, messages }) => {
      let urls = messages[0];
      let imports = messages[1];

      // TODO(michael-ciniawsky) revisit
      // Ensure to cleanup/reset messages
      // during recursive resolving of imports
      messages.length = 0;

      // <img src="./file.png">
      // => import HTML__URL__${idx} from './file.png';
      if (urls) {
        urls = Object.keys(urls)
          .map((url) => `import ${url} from '${urls[url]}';`)
          .join('\n');
      }
      // <import src="./file.html">
      // => import HTML__IMPORT__${idx} from './file.html';
      if (imports) {
        imports = Object.keys(imports)
          .map((i) => `import ${i} from '${imports[i]}';`)
          .join('\n');
      }

      html = options.template
        ? `function (${template}) { return \`${html}\`; }`
        : `\`${html}\``;

      const result = [
        urls ? `// HTML URLs\n${urls}\n` : false,
        imports ? `// HTML Imports\n${imports}\n` : false,
        `// HTML\nexport default ${html}`,
      ]
        .filter(Boolean)
        .join('\n');

      cb(null, result);

      return null;
    })
    .catch((err) => {
      cb(new LoaderError(err));

      return null;
    });
}
