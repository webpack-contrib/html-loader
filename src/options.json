{
  "type": "object",
  "definitions": {
    "Attribute": {
      "type": "object",
      "properties": {
        "tag": {
          "type": "string",
          "minLength": 1
        },
        "attribute": {
          "type": "string",
          "minLength": 1
        },
        "type": {
          "enum": ["src", "srcset"]
        },
        "filter": {
          "instanceof": "Function"
        }
      },
      "required": ["tag", "attribute", "type"],
      "additionalProperties": false
    },
    "AttributeList": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/Attribute"
      },
      "minItems": 1,
      "uniqueItems": true
    }
  },
  "properties": {
    "preprocessor": {
      "instanceof": "Function"
    },
    "attributes": {
      "anyOf": [
        { "type": "boolean" },
        {
          "type": "object",
          "properties": {
            "list": {
              "$ref": "#/definitions/AttributeList"
            },
            "urlFilter": {
              "instanceof": "Function"
            },
            "root": {
              "type": "string"
            }
          },
          "additionalProperties": false
        }
      ]
    },
    "minimize": {
      "anyOf": [{ "type": "boolean" }, { "type": "object" }]
    },
    "esModule": {
      "type": "boolean"
    }
  },
  "additionalProperties": false
}
