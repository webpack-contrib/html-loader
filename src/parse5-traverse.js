const traverse = (root, callback) => {
  const visit = (node, parent) => {
    let res;

    if (callback) {
      res = callback(node, parent);
    }

    let { childNodes } = node;

    // in case a <template> tag is in the middle of the HTML: https://github.com/JPeer264/node-rcs-core/issues/58
    if (node.content && Array.isArray(node.content.childNodes)) {
      ({ childNodes } = node.content);
    }

    if (res !== false && Array.isArray(childNodes) && childNodes.length >= 0) {
      childNodes.forEach((child) => {
        visit(child, node);
      });
    }
  };

  visit(root, null);
};

module.exports = traverse;
