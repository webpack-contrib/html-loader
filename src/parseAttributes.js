import Parser from 'fastparse';

function processMatch(match, strUntilValue, name, value, index) {
  if (!this.isRelevantTagAttr(this.currentTag, name)) {
    return;
  }

  this.results.push({
    start: index + strUntilValue.length,
    length: value.length,
    value,
  });
}

const parser = new Parser({
  outside: {
    '<!--.*?-->': true,
    '<![CDATA[.*?]]>': true,
    '<[!\\?].*?>': true,
    '</[^>]+>': true,
    '<([a-zA-Z\\-:]+)\\s*': function matchTag(match, tagName) {
      this.currentTag = tagName;

      return 'inside';
    },
  },
  inside: {
    // eat up whitespace
    '\\s+': true,
    // end of attributes
    '>': 'outside',
    '(([0-9a-zA-Z\\-:]+)\\s*=\\s*")([^"]*)"': processMatch,
    "(([0-9a-zA-Z\\-:]+)\\s*=\\s*')([^']*)'": processMatch,
    '(([0-9a-zA-Z\\-:]+)\\s*=\\s*)([^\\s>]+)': processMatch,
  },
});

export default function parseAttributes(html, isRelevantTagAttr) {
  return parser.parse('outside', html, {
    currentTag: null,
    results: [],
    isRelevantTagAttr,
  }).results;
}
