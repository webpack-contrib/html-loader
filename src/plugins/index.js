import sourcesPlugin from "./sources-plugin";
import minimizerPlugin from "./minimizer-plugin";

export { sourcesPlugin, minimizerPlugin };
