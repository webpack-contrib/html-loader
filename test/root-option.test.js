import loader from '../src';
import { GET_URL_CODE } from '../src/constants';

describe("'root' option", () => {
  it('should not translate root-relative urls by default', () => {
    const result = loader.call(
      { mode: 'development' },
      'Text <img src="/image.png">'
    );

    expect(result).toBe('module.exports = "Text <img src=\\"/image.png\\">";');
  });

  it('should work', () => {
    const result = loader.call(
      {
        mode: 'development',
        query: '?root=/test',
      },
      'Text <img src="/image.png">'
    );

    expect(result).toBe(
      `${GET_URL_CODE}module.exports = "Text <img src=\\"" + __url__(require("/test/image.png")) + "\\">";`
    );
  });
});
