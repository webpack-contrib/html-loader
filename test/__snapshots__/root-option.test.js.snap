// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'root' option should not translate root-relative urls by default: errors 1`] = `Array []`;

exports[`'root' option should not translate root-relative urls by default: module 1`] = `
"// Imports
var ___HTML_LOADER_GET_URL_IMPORT___ = require(\\"../../src/runtime/getUrl.js\\");
var ___HTML_LOADER_IDENT_0___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_1___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_2___ = require(\\"aliasImg\\");
var ___HTML_LOADER_IDENT_3___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_4___ = require(\\"./icons.svg\\");
var ___HTML_LOADER_IDENT_5___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_6___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_7___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_8___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_9___ = require(\\"./image.png\\");
// Exports
module.exports = \\"<h1>My First Heading</h1>\\\\n<p>My first paragraph.</p>\\\\n<h2>An Unordered HTML List</h2>\\\\n\\\\n<ul>\\\\n  <li>Coffee</li>\\\\n  <li>Tea</li>\\\\n  <li>Milk</li>\\\\n</ul>\\\\n\\\\n<h2>An Ordered HTML List</h2>\\\\n\\\\n<ol>\\\\n  <li>Coffee</li>\\\\n  <li>Tea</li>\\\\n  <li>Milk</li>\\\\n</ol>\\\\n\\\\n<script>console.log({\\\\\\"json\\\\\\": \\\\\\"with \\\\\\\\\\\\\\"quotes\\\\\\\\\\\\\\" in value\\\\\\"})</script>\\\\n\\\\n<!-- comment -->\\\\n\\\\n<div>Foo</div>\\\\n\\\\n\\\\n<div>BAR</div>\\\\n\\\\n<input type=\\\\\\"text\\\\\\" />\\\\n\\\\n<script>   console.log(1 + 2 + \`\${3 + 3}\`)   </script>\\\\n\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_0___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_1___) + \\"\\\\\\">\\\\n<img src=\\\\\\"/image.png\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_2___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_3___) + \\"\\\\\\">\\\\n\\\\n<script src=\\\\\\"script.js\\\\\\"></script>\\\\n\\\\n<custom-element custom-src=\\\\\\"image.png\\\\\\" />\\\\n<custom-img custom-src=\\\\\\"image.png\\\\\\"/></custom-element>\\\\n\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_4___) + \\"#hash\\\\\\">\\\\n\\\\n<a href=\\\\\\"mailto:username@exampledomain.com\\\\\\"></a>\\\\n\\\\n<img srcset=\\\\\\"image.png 480w, image.png 800w\\\\\\" sizes=\\\\\\"(max-width: 600px) 480px, 800px\\\\\\" src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_5___) + \\"\\\\\\" alt=\\\\\\"Elva dressed as a fairy\\\\\\">\\\\n<img srcset=\\\\\\"   image.png   480w   ,   image.png   800w   \\\\\\" sizes=\\\\\\"(max-width: 600px) 480px, 800px\\\\\\" src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_6___) + \\"\\\\\\" alt=\\\\\\"Elva dressed as a fairy\\\\\\">\\\\n\\\\n<picture>\\\\n  <source media=\\\\\\"(min-width: 650px)\\\\\\" srcset=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_7___) + \\"\\\\\\">\\\\n  <source media=\\\\\\"(min-width: 465px)\\\\\\" srcset=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_8___) + \\"\\\\\\">\\\\n  <img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_9___) + \\"\\\\\\" alt=\\\\\\"Flowers\\\\\\" style=\\\\\\"width:auto;\\\\\\">\\\\n</picture>\\\\n\\";"
`;

exports[`'root' option should not translate root-relative urls by default: result 1`] = `
"<h1>My First Heading</h1>
<p>My first paragraph.</p>
<h2>An Unordered HTML List</h2>

<ul>
  <li>Coffee</li>
  <li>Tea</li>
  <li>Milk</li>
</ul>

<h2>An Ordered HTML List</h2>

<ol>
  <li>Coffee</li>
  <li>Tea</li>
  <li>Milk</li>
</ol>

<script>console.log({\\"json\\": \\"with \\\\\\"quotes\\\\\\" in value\\"})</script>

<!-- comment -->

<div>Foo</div>


<div>BAR</div>

<input type=\\"text\\" />

<script>   console.log(1 + 2 + \`\${3 + 3}\`)   </script>

<img src=\\"/webpack/public/path/image.png\\">
<img src=\\"/webpack/public/path/image.png\\">
<img src=\\"/image.png\\">
<img src=\\"/webpack/public/path/alias-image.png\\">
<img src=\\"\\">
<img src=\\"/webpack/public/path/image.png\\">

<script src=\\"script.js\\"></script>

<custom-element custom-src=\\"image.png\\" />
<custom-img custom-src=\\"image.png\\"/></custom-element>

<img src=\\"/webpack/public/path/icons.svg#hash\\">

<a href=\\"mailto:username@exampledomain.com\\"></a>

<img srcset=\\"image.png 480w, image.png 800w\\" sizes=\\"(max-width: 600px) 480px, 800px\\" src=\\"/webpack/public/path/image.png\\" alt=\\"Elva dressed as a fairy\\">
<img srcset=\\"   image.png   480w   ,   image.png   800w   \\" sizes=\\"(max-width: 600px) 480px, 800px\\" src=\\"/webpack/public/path/image.png\\" alt=\\"Elva dressed as a fairy\\">

<picture>
  <source media=\\"(min-width: 650px)\\" srcset=\\"/webpack/public/path/image.png\\">
  <source media=\\"(min-width: 465px)\\" srcset=\\"/webpack/public/path/image.png\\">
  <img src=\\"/webpack/public/path/image.png\\" alt=\\"Flowers\\" style=\\"width:auto;\\">
</picture>
"
`;

exports[`'root' option should not translate root-relative urls by default: warnings 1`] = `Array []`;

exports[`'root' option should work: errors 1`] = `Array []`;

exports[`'root' option should work: module 1`] = `
"// Imports
var ___HTML_LOADER_GET_URL_IMPORT___ = require(\\"../../src/runtime/getUrl.js\\");
var ___HTML_LOADER_IDENT_0___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_1___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_2___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_3___ = require(\\"aliasImg\\");
var ___HTML_LOADER_IDENT_4___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_5___ = require(\\"./icons.svg\\");
var ___HTML_LOADER_IDENT_6___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_7___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_8___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_9___ = require(\\"./image.png\\");
var ___HTML_LOADER_IDENT_10___ = require(\\"./image.png\\");
// Exports
module.exports = \\"<h1>My First Heading</h1>\\\\n<p>My first paragraph.</p>\\\\n<h2>An Unordered HTML List</h2>\\\\n\\\\n<ul>\\\\n  <li>Coffee</li>\\\\n  <li>Tea</li>\\\\n  <li>Milk</li>\\\\n</ul>\\\\n\\\\n<h2>An Ordered HTML List</h2>\\\\n\\\\n<ol>\\\\n  <li>Coffee</li>\\\\n  <li>Tea</li>\\\\n  <li>Milk</li>\\\\n</ol>\\\\n\\\\n<script>console.log({\\\\\\"json\\\\\\": \\\\\\"with \\\\\\\\\\\\\\"quotes\\\\\\\\\\\\\\" in value\\\\\\"})</script>\\\\n\\\\n<!-- comment -->\\\\n\\\\n<div>Foo</div>\\\\n\\\\n\\\\n<div>BAR</div>\\\\n\\\\n<input type=\\\\\\"text\\\\\\" />\\\\n\\\\n<script>   console.log(1 + 2 + \`\${3 + 3}\`)   </script>\\\\n\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_0___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_1___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_2___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_3___) + \\"\\\\\\">\\\\n<img src=\\\\\\"\\\\\\">\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_4___) + \\"\\\\\\">\\\\n\\\\n<script src=\\\\\\"script.js\\\\\\"></script>\\\\n\\\\n<custom-element custom-src=\\\\\\"image.png\\\\\\" />\\\\n<custom-img custom-src=\\\\\\"image.png\\\\\\"/></custom-element>\\\\n\\\\n<img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_5___) + \\"#hash\\\\\\">\\\\n\\\\n<a href=\\\\\\"mailto:username@exampledomain.com\\\\\\"></a>\\\\n\\\\n<img srcset=\\\\\\"image.png 480w, image.png 800w\\\\\\" sizes=\\\\\\"(max-width: 600px) 480px, 800px\\\\\\" src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_6___) + \\"\\\\\\" alt=\\\\\\"Elva dressed as a fairy\\\\\\">\\\\n<img srcset=\\\\\\"   image.png   480w   ,   image.png   800w   \\\\\\" sizes=\\\\\\"(max-width: 600px) 480px, 800px\\\\\\" src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_7___) + \\"\\\\\\" alt=\\\\\\"Elva dressed as a fairy\\\\\\">\\\\n\\\\n<picture>\\\\n  <source media=\\\\\\"(min-width: 650px)\\\\\\" srcset=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_8___) + \\"\\\\\\">\\\\n  <source media=\\\\\\"(min-width: 465px)\\\\\\" srcset=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_9___) + \\"\\\\\\">\\\\n  <img src=\\\\\\"\\" + ___HTML_LOADER_GET_URL_IMPORT___(___HTML_LOADER_IDENT_10___) + \\"\\\\\\" alt=\\\\\\"Flowers\\\\\\" style=\\\\\\"width:auto;\\\\\\">\\\\n</picture>\\\\n\\";"
`;

exports[`'root' option should work: result 1`] = `
"<h1>My First Heading</h1>
<p>My first paragraph.</p>
<h2>An Unordered HTML List</h2>

<ul>
  <li>Coffee</li>
  <li>Tea</li>
  <li>Milk</li>
</ul>

<h2>An Ordered HTML List</h2>

<ol>
  <li>Coffee</li>
  <li>Tea</li>
  <li>Milk</li>
</ol>

<script>console.log({\\"json\\": \\"with \\\\\\"quotes\\\\\\" in value\\"})</script>

<!-- comment -->

<div>Foo</div>


<div>BAR</div>

<input type=\\"text\\" />

<script>   console.log(1 + 2 + \`\${3 + 3}\`)   </script>

<img src=\\"/webpack/public/path/image.png\\">
<img src=\\"/webpack/public/path/image.png\\">
<img src=\\"/webpack/public/path/image.png\\">
<img src=\\"/webpack/public/path/alias-image.png\\">
<img src=\\"\\">
<img src=\\"/webpack/public/path/image.png\\">

<script src=\\"script.js\\"></script>

<custom-element custom-src=\\"image.png\\" />
<custom-img custom-src=\\"image.png\\"/></custom-element>

<img src=\\"/webpack/public/path/icons.svg#hash\\">

<a href=\\"mailto:username@exampledomain.com\\"></a>

<img srcset=\\"image.png 480w, image.png 800w\\" sizes=\\"(max-width: 600px) 480px, 800px\\" src=\\"/webpack/public/path/image.png\\" alt=\\"Elva dressed as a fairy\\">
<img srcset=\\"   image.png   480w   ,   image.png   800w   \\" sizes=\\"(max-width: 600px) 480px, 800px\\" src=\\"/webpack/public/path/image.png\\" alt=\\"Elva dressed as a fairy\\">

<picture>
  <source media=\\"(min-width: 650px)\\" srcset=\\"/webpack/public/path/image.png\\">
  <source media=\\"(min-width: 465px)\\" srcset=\\"/webpack/public/path/image.png\\">
  <img src=\\"/webpack/public/path/image.png\\" alt=\\"Flowers\\" style=\\"width:auto;\\">
</picture>
"
`;

exports[`'root' option should work: warnings 1`] = `Array []`;
